### DISCUSSION TITLE
Premature commit

### DISCUSSION THREAD

**Discussion by heyarnold1** [sqlalchemy/sqlalchemy#6677] [Usage Questions]
We experience an intermittent sqlalchemy.exc.IntegrityError error that only occurs in our Github actions testing environment.  We have not been able to recreate it locally or in production on AWS.  However we can prevent it with a small delay, see below.

Sqlalchemy appears in the failing cases to try and prematurely commit an update.  I suspect the update is being committed as part of the previous transaction.  I cannot provide code that will easily recreate this issue, but I hope this code gives an indication of what is happening.

I am wandering if anyone else has experience of this?

```py
class Table1(Base):
    id = Column(Integer, primary_key=True)
    name = Column(String, nullable=False)

class Table2(Base):
    id = Column(Integer, primary_key=True)
    another_id = Column(Integer, primary_key=True)

class Table3(Base):
    id = Column(Integer, primary_key=True)
    another_id = Column(Integer, primary_key=True)


    __table_args__ = (
        ForeignKeyConstraint(
            [id, another_id],
            [Table2.id, Table2.another_id],
        ),
    )



db.query(models.Table1).filter(id==10).update({"name": 'something'}) # this is not important
db.commit()

#add sleep or db query here to prevent error

db.query(models.Table2).filter(id==1).update({"another_id": 2}) # sqlalchemy.exc.IntegrityError
db.query(models.Table3).filter(id==1).update({"another_id": 2})
db.commit()
```

**Example Error** 
``sqlalchemy.exc.IntegrityError: (psycopg2.errors.ForeignKeyViolation) update or delete on table "Table2" violates foreign key constraint "fk__tabel1__table2" on table "Table3"``

**Versions.**
 - Python: 3.9
 - SQLAlchemy: 1.3.24
 - Database: Postgres 10.15
 - DBAPI: psycopg2 2.9.1

**Comment by zzzeek** [comment]
the error message you're showing is not specifically from a "commit", it's from a [flush](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#flushing).   the session will autoflush whenever you run a query, so those lines where you do "db.query(..).update()", those will autoflush by default.   you can turn off autoflush either session-wise or for a specific block as indicated in the linked document.

https://docs.sqlalchemy.org/en/14/orm/session_basics.html#flushing
